File: TypeScript/helper/grid.ts
// Used in:
//  Queen Attack
//  Word Search

export enum Direction {
    N,
    E,
    S,
    W,
    NE,
    SE,
    SW,
    NW,
};

export interface Offset {
    x: number;
    y: number;
}

export interface OffsetData {
    direction: Direction;
    value: any;
}

export const Offsets: Record<Direction, Offset> = Object.freeze({
    [Direction.N]: { x: -1, y: 0 },
    [Direction.S]: { x: 1, y: 0 },
    [Direction.E]: { x: 0, y: 1 },
    [Direction.W]: { x: 0, y: -1 },
    [Direction.NE]: { x: -1, y: 1 },
    [Direction.SE]: { x: 1, y: 1 },
    [Direction.SW]: { x: 1, y: -1 },
    [Direction.NW]: { x: -1, y: -1 },
});

/**
 * Represents a 2D grid.
 */
export class Grid<T> {
    data: T[][];

    /**
     * Creates a new Grid instance.
     * @param {number} rows - The number of rows in the grid.
     * @param {number} columns - The number of columns in the grid.
     * @param {T} fillValue - The value to fill the grid cells with. undefined by default
     */
    constructor(rows: number, columns: number, fillValue: T | undefined = undefined) {
        this.data = [];

        for(let i = 0; i < rows; i++) {
            let col: T[] = [];

            for(let j = 0; j < columns; j++) {
                col.push(fillValue as T);
            };

            this.data.push(col);
        };
    };

    /**
     * Creates a Grid instance from existing data.
     * @param {Array<Array<T>>} data - The 2D array representing the grid data.
     * @returns {Grid<T>} - A new Grid instance.
     */
     static fromData<T>(data: T[][]): Grid<T> {
        let grid = new Grid<T>(data.length, data[0].length);

        for (let i = 0; i < grid.data.length; i++) {
            for (let j = 0; j < grid.data[i].length; j++) {
                grid.data[i][j] = data[i][j];
            }
        }

        return grid;
    }

    /**
     * Gets the size of the grid.
     * @returns {{ rows: number, columns: number }} - The size of the grid.
     */
    get size() {
        return { rows: this.data.length, columns: this.data[0].length };
    }

    /**
     * Determines if a cell is in the grid's dimensions.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @returns {boolean} - Returns true if the x y coordinate is in the grid.
     */
    isValidCoordinate(x: number, y: number): boolean {
        return x >= 0 && x < this.size.rows && y >= 0 && y < this.size.columns;
    }

    /**
     * Gets the value at the specified cell.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @returns {T} - The value at the specified cell.
     * @throws {Error} - Throws an error if the coordinates are out of range.
     */
    cell(x: number, y: number): T {
        if (this.isValidCoordinate(x, y)) {
            return this.data[x][y];
        } 
        else {
            throw new Error("Invalid coordinates");
        }
    }

    /**
     * Iterates through each element in the grid and applies the given callback function.
     * 
     * @param callback - The function to call for each element, which receives the value, row index, and column index.
     */
    forEach(callback: (value: T, row: number, col: number) => void): void {
        for (let row = 0; row < this.data.length; row++) {
            for (let col = 0; col < this.data[row].length; col++) {
                callback(this.data[row][col], row, col);
            }
        }
    }

    /**
     * Flattens the 2D grid into a 1D array.
     * 
     * @returns A single array containing all the elements of the grid.
     */
    flat(): T[] {
        return this.data.reduce((acc, row) => acc.concat(row), []);
    }

    /**
     * Sets the value at the specified cell.
     * 
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @param {T} fillValue - The value to set at the specified cell.
     * @throws {Error} - Throws an error if the coordinates are out of range.
     */
    fillCell(x: number, y: number, fillValue: T): void {
        if (this.isValidCoordinate(x, y)) {
            this.data[x][y] = fillValue;
        } 
        else {
            throw new Error("Invalid coordinates");
        }
    }

    /**
     * Gets data from with an offset relative to a position.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @param {{ x: number, y: number }} offset - An object representing the x y offset.
     * @returns {T | undefined} - The value in a nearby cell, or undefined if out of bounds.
     */
    getOffsetData(x: number, y: number, offset: Offset): T | undefined {
        const newX = x + offset.x;
        const newY = y + offset.y;

        if (this.isValidCoordinate(newX, newY)) {
            return this.data[newX][newY];
        }

        return undefined;
    }

    /**
     * Gets all data from an offset direction relative to a position.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @param {Offset} offset - An object representing the x y offset.
     * @returns {Array<T>} - The values in a direction.
     */
    getOffsetDataLoop(x: number, y: number, offset: Offset): T[] {
        let values: T[] = [];
        let counter = 1;

        let value = this.getOffsetData(x, y, { x: 0, y: 0 });
        while (value !== undefined) {
            values.push(value);

            value = this.getOffsetData(x, y, { x: offset.x * counter, y: offset.y * counter });
            counter++;
        }

        return values;
    }

    /**
     * Gets all data from a row in the grid.
     * @param {number} n - The row index of the grid.
     * @returns {number} - The values from a row in the grid.
     * @throws {Error} - Throws an error If the row index is out of bounds.
     */
    row(n: number): T[] {
        if (n >= 0 && n < this.size.rows) {
            return this.data[n];
        } 

        else {
            throw new Error("Invalid row index");
        }
    }

    /**
     * Gets all data from a column in the grid.
     * @param {number} n - The column index of the grid.
     * @returns {array} - The values from a column in the grid.
     * @throws {Error} - Throws an error If the column index is out of bounds.
     */
    column(n: number): T[] {
        if (n >= 0 && n < this.size.columns) {
            return this.data.map(row => row[n]);
        } 

        else {
            throw new Error("Invalid column index");
        }
    }

    /**
     * Gets all data from a cardinal direction (N, E, S, W) including the specified cell.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @param {Direction} direction - The cardinal direction.
     * @returns {Array} - The values from a cardinal direction.
     */
    cardinal(x: number, y: number, direction: Direction): T[] {
        const offset = Offsets[direction];

        if (offset) {
            return this.getOffsetDataLoop(x, y, offset);
        } 
        
        else {
            throw new Error("Cardinal direction not recognized.");
        }
    }

    /**
     * Gets all data from a ordinal direction (NE, SE, SW, NW) including the specified cell.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @param {Direction} direction - The ordinal direction.
     * @returns {Array} - The values from a ordinal direction.
     */
    ordinal(x: number, y: number, direction: Direction): T[] {
        const offset = Offsets[direction];

        if (offset) {
            return this.getOffsetDataLoop(x, y, offset);
        } 
        
        else {
            throw new Error("Cardinal direction not recognized.");
        }
    }

    /**
     * Gets all data from all cardinal directions (N, E, S, W) including the specified cell.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @returns {Array} - An array of objects from all cardinal directions.
     */
    cardinals(x: number, y: number) {
        return [
            { direction:Direction.N, values:this.cardinal(x, y, Direction.N) },
            { direction:Direction.S, values:this.cardinal(x, y, Direction.S) },
            { direction:Direction.E, values:this.cardinal(x, y, Direction.E) },
            { direction:Direction.W, values:this.cardinal(x, y, Direction.W) },
        ];
    }

    /**
     * Gets all data from all ordinal directions (NE, SE, SW, NW) including the specified cell.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @returns {Array} - An array of objects from all ordinal directions.
     */
    ordinals(x: number, y: number) {
        return [
            { direction:Direction.NE, values:this.ordinal(x, y, Direction.NE) },
            { direction:Direction.SE, values:this.ordinal(x, y, Direction.SE) },
            { direction:Direction.SW, values:this.ordinal(x, y, Direction.SW) },
            { direction:Direction.NW, values:this.ordinal(x, y, Direction.NW) },
        ];
    }

    /**
     * Gets all data from all cardinal & ordinal directions (N, E, S, W, NE, SE, SW, NW) including the specified cell.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @returns {Array} - An array of objects from all cardinal & ordinal directions.
     */
    cardinalsAndOrdinals(x: number, y: number) {
        return [
            { direction:Direction.N, values:this.cardinal(x, y, Direction.N) },
            { direction:Direction.S, values:this.cardinal(x, y, Direction.S) },
            { direction:Direction.E, values:this.cardinal(x, y, Direction.E) },
            { direction:Direction.W, values:this.cardinal(x, y, Direction.W) },
            { direction:Direction.NE, values:this.ordinal(x, y, Direction.NE) },
            { direction:Direction.SE, values:this.ordinal(x, y, Direction.SE) },
            { direction:Direction.SW, values:this.ordinal(x, y, Direction.SW) },
            { direction:Direction.NW, values:this.ordinal(x, y, Direction.NW) },
        ];
    }

    /**
     * Returns data for the neighbors of a given cell in the grid.
     * @param {number} x - The row index of the cell.
     * @param {number} y - The column index of the cell.
     * @param {...string} directions - Optional. Directions to search for neighbors. 
     *                                 If not provided, searches in all directions.
     * @returns {Array} - An array containing data for the neighbors in specified directions.
     */
    getNeighbors(x: number, y: number, ...directions: Direction[]): OffsetData[] {
        const data: OffsetData[] = [];

        if (directions.length === 0) {
            directions = [Direction.N, Direction.E, Direction.S, Direction.W, Direction.NE, Direction.SE, Direction.SW, Direction.NW];
        }

        for (const dir of directions) {
            const value = this.getOffsetData(x, y, Offsets[dir]);
            data.push({ direction: dir, value });
        }
        return data;
    }

    /**
     * Returns a string representation of the grid For easy printing.
     * @returns {string} - A string representation of the grid.
     */
    toString(): string {
        return this.data.map(row => row.join(" ")).join("\n");
    }
}

File: C#/_Helper/Grid.cs
using System;
using System.Collections.Generic;
using System.Linq;

/// <summary>
/// A class for 2D grids. 
/// 
/// Used in
///     - Minesweeper
///     - Queen Attack
///     - Word Search
/// 
/// It is 0 based and accessed by row then column
/// </summary>

namespace _Helper
{
    public enum Direction
    {
        N,
        E,
        S,
        W,
        NE,
        SE,
        SW,
        NW,
    }

    public class Offset
    {
        public int X { get; set; }
        public int Y { get; set; }
    }

    public class OffsetData<T>
    {
        public Direction Direction { get; set; }
        public T? Value { get; set; }
    }

    public class OffsetDataArray<T>
    {
        public Direction Direction { get; set; }
        public T[]? Values { get; set; }
    }

    public static class Offsets
    {
        public static readonly Dictionary<Direction, Offset> OffsetDictionary = new Dictionary<Direction, Offset>
        {
            { Direction.N, new Offset { X = -1, Y = 0 } },
            { Direction.S, new Offset { X = 1, Y = 0 } },
            { Direction.E, new Offset { X = 0, Y = 1 } },
            { Direction.W, new Offset { X = 0, Y = -1 } },
            { Direction.NE, new Offset { X = -1, Y = 1 } },
            { Direction.SE, new Offset { X = 1, Y = 1 } },
            { Direction.SW, new Offset { X = 1, Y = -1 } },
            { Direction.NW, new Offset { X = -1, Y = -1 } }
        };
    }

    public class Grid<T>
    {
        public List<List<T>> _data = [];

        public (int, int) Size { get; }

        public Grid(int rows, int columns, T fillValue = default)
        {
            for (int i = 0; i < rows; i++)
            {
                var col = new List<T>();

                for (int j = 0; j < columns; j++)
                {
                    col.Add(fillValue);
                };

                this._data.Add(col);
            };

            this.Size = (_data.Count, _data.First().Count);
        }

        public static Grid<T> FromData(T[][] data)
        {
            Grid<T> grid = new Grid<T>(data.Length, data[0].Length);

            for (int i = 0; i < grid._data.Count; i++)
            {
                for (int j = 0; j < grid._data[i].Count; j++)
                {
                    grid._data[i][j] = data[i][j];
                }
            }

            return grid;
        }

        public static Grid<string> FromStringArray(string[] data)
        {
            string[][] stringData = new string[data.Length][];

            for (int i = 0; i < data.Length; i++)
            {
                stringData[i] = new string[data[i].Length];
                for (int j = 0; j < data[i].Length; j++)
                {
                    stringData[i][j] = data[i][j].ToString();
                }
            }

            return Grid<string>.FromData(stringData);
        }

        public bool IsValidCoordinate(int x, int y) => x >= 0 && x < this.Size.Item1 && y >= 0 && y < this.Size.Item2;

        public T Cell(int x, int y)
        {
            if(this.IsValidCoordinate(x, y) == false)
            {
                throw new ArgumentOutOfRangeException("Co-ordinate is not in the grid");
            }

            return _data[x][y];
        }

        public void FillCell(int x, int y, T value)
        {
            if (IsValidCoordinate(x, y))
            {
                _data[x][y] = value;
            }
            else
            {
                throw new Exception("Co-ordinate is not in the grid");
            }
        }

        public void ForEach(Action<T, int, int> callback)
        {
            for (int row = 0; row < _data.Count; row++)
            {
                for (int col = 0; col < _data[0].Count; col++)
                {
                    callback(_data[row][col], row, col);
                }
            }
        }

        public T[] Flat()
        {
            List<T> flatList = new List<T>();

            foreach(var row in _data)
            {
                flatList.AddRange(row);
            }

            return flatList.ToArray();
        }

        public T? GetOffsetData(int x, int y, Offset offset)
        {
            int newX = x + offset.X;
            int newY = y + offset.Y;

            if (this.IsValidCoordinate(newX, newY))
            {
                return _data[newX][newY];
            }

            return default(T);
        }

        public T[] GetOffsetDataLoop(int x, int y, Offset offset)
        {
            List<T> values = new List<T>();
            int counter = 1;

            T value = GetOffsetData(x, y, new Offset { X = 0, Y = 0 });
            while (!EqualityComparer<T>.Default.Equals(value, default(T)))
            {
                values.Add(value);

                value = GetOffsetData(x, y, new Offset { X = offset.X * counter, Y = offset.Y * counter });
                counter++;
            }

            return values.ToArray();
        }

        public T[] Row(int n)
        {
            if (n >= 0 && n < _data.Count)
            {
                return _data[n].ToArray();
            }
            else
            {
                throw new Exception("Invalid row index");
            }
        }

        public T[] Column(int n)
        {
            if (n >= 0 && n < _data[0].Count)
            {
                List<T> column = new List<T>();

                foreach (var row in _data)
                {
                    column.Add(row[n]);
                }

                return column.ToArray();
            }
            else
            {
                throw new Exception("Invalid column index");
            }
        }

        public T[] Cardinal(int x, int y, Direction direction)
        {
            Offset offset = Offsets.OffsetDictionary[direction];

            if (offset != null)
            {
                return GetOffsetDataLoop(x, y, offset);
            }
            else
            {
                throw new Exception("Cardinal direction not recognized.");
            }
        }

        public T[] Ordinal(int x, int y, Direction direction)
        {
            Offset offset = Offsets.OffsetDictionary[direction];

            if (offset != null)
            {
                return GetOffsetDataLoop(x, y, offset);
            }
            else
            {
                throw new Exception("Ordinal direction not recognized.");
            }
        }

        public T[] Cardinals(int x, int y)
        {
            return new T[][]
            {
                Cardinal(x, y, Direction.N),
                Cardinal(x, y, Direction.S),
                Cardinal(x, y, Direction.E),
                Cardinal(x, y, Direction.W)
            }.SelectMany(a => a).ToArray();
        }

        public T[] Ordinals(int x, int y)
        {
            return new T[][]
            {
                Ordinal(x, y, Direction.NE),
                Ordinal(x, y, Direction.SE),
                Ordinal(x, y, Direction.SW),
                Ordinal(x, y, Direction.NW)
            }.SelectMany(a => a).ToArray();
        }

        public OffsetDataArray<T>[] CardinalsAndOrdinals(int x, int y)
        {
            List<OffsetDataArray<T>> dataList =
            [
                new OffsetDataArray<T> { Direction = Direction.N, Values = Cardinal(x, y, Direction.N) },
                new OffsetDataArray<T> { Direction = Direction.S, Values = Cardinal(x, y, Direction.S) },
                new OffsetDataArray<T> { Direction = Direction.E, Values = Cardinal(x, y, Direction.E) },
                new OffsetDataArray<T> { Direction = Direction.W, Values = Cardinal(x, y, Direction.W) },
                new OffsetDataArray<T> { Direction = Direction.NE, Values = Ordinal(x, y, Direction.NE) },
                new OffsetDataArray<T> { Direction = Direction.SE, Values = Ordinal(x, y, Direction.SE) },
                new OffsetDataArray<T> { Direction = Direction.SW, Values = Ordinal(x, y, Direction.SW) },
                new OffsetDataArray<T> { Direction = Direction.NW, Values = Ordinal(x, y, Direction.NW) },
            ];

            return dataList.ToArray();
        }

        public OffsetData<T>[] GetNeighbors(int x, int y, params Direction[] directions)
        {
            List<OffsetData<T>> dataList = new List<OffsetData<T>>();

            if (directions.Length == 0)
            {
                directions = new Direction[] { Direction.N, Direction.E, Direction.S, Direction.W, Direction.NE, Direction.SE, Direction.SW, Direction.NW };
            }

            foreach (Direction dir in directions)
            {
                T? value = GetOffsetData(x, y, Offsets.OffsetDictionary[dir]);
                dataList.Add(new OffsetData<T> { Direction = dir, Value = value });
            }

            return dataList.ToArray();
        }

        public override string ToString() => string.Join("\n", _data.Select(row => string.Join(" ", row)));
    }
}